### DISCUSSION TITLE
Multi step process to allow scanning and testing artifacts before publishing

### DISCUSSION THREAD

**Discussion by MadsRC** [goreleaser/goreleaser#4879] [Q&A] [answered]
I am looking to have a release and publishing process that allows me effectively run a suite of tests and scans on artifacts to make sure they are behaving as expected and only then publish them. The key aspect here is that the artifacts that are being published must be the ones that were tested and scanned.

Does anyone have a similar release and publishing process where they use GoReleaser (OSS/Free or Pro)?

Looking at the GoReleaser docs, the only way I see this is possible is with the [--split](https://goreleaser.com/customization/partial/) argument - The main use-case for this seem to be speeding up builds, which is not what I'm looking for though.

An alternative for me is to just run GoRelease snapshot, run scans and integration tests on the snapshot release and then, if all is good, run GoRelease again to create a proper release. While this works, it isn't ideal as I'm not actually testing the released/published artifacts.

Any input would be much appreciated ;)

**Comment by caarlos0** [comment, marked as answer]
hey!

I have in my plans to add some sort of pre-publish hook, in which you could run anything with the binaries/packages/etc.

I haven't started working on it yet, or have a complete idea of how it should look like, but probably something like:


```yml
pre_publish_checks:
  - ids: [foo]
    artifacts: [ 'binary', 'archive' ]
    run: './integration-test.sh {{.Path}}'
    fail_on_error: true
```

would love to hear your opinions and ideas on it!

**Reply by MadsRC** [reply]
That would be a nice addition, I think that would allow me to implement my use-case!

**Reply by caarlos0** [reply]
added to goreleaser pro, will be on the next release
